community/kernel/src/main/java/org/neo4j/storageengine/api/StorageReader.java

```diff
@@ -46,29 +50,169 @@
 import org.neo4j.kernel.impl.api.DegreeVisitor;
 import org.neo4j.kernel.impl.api.RelationshipVisitor;
 import org.neo4j.kernel.impl.api.store.RelationshipIterator;
+import org.neo4j.kernel.impl.locking.Lock;
+import org.neo4j.kernel.impl.store.RecordCursors;
 import org.neo4j.register.Register.DoubleLongRegister;
+import org.neo4j.storageengine.api.schema.IndexReader;
+import org.neo4j.storageengine.api.schema.LabelScanReader;
 import org.neo4j.storageengine.api.schema.PopulationProgress;
 
 /**
- * Abstraction for reading committed data from {@link StorageEngine store}.
+ * Abstraction for accessing data from a {@link StorageEngine}.
+ * <p>
+ * A {@link StorageReader} must be initialized before use in a
+ * {@link #beginTransaction(TransactionalDependencies) transaction}/{@link #beginStatement() statement}. After use there must
+ * be a call to end {@link #endStatement() statement}/{@link #endTransaction() transaction}.
+ * After released the reader can be acquired again.
+ * <p>
+ * Creating and closing {@link StorageReader} can be somewhat costly, so there are benefits keeping these readers open
+ * during a longer period of time, with the assumption that it's still one thread at a time using each.
  */
-public interface StoreReadLayer
+public interface StorageReader extends AutoCloseable, Read, ExplicitIndexRead, SchemaRead, CursorFactory
```

Comment:
also i think its a step in the wrong direction to create a monster interface like StorageReader that will gonna do everything. This was exactly the driver for a couple of previous refactorings to split that work into separate functionality isolated classes which i think is a good thing, this brings us back.